src/Microsoft.Diagnostics.DebugServices.Implementation/TargetFromDataReader.cs

```diff
@@ -33,36 +33,39 @@ public TargetFromDataReader(IDataReader dataReader, OSPlatform targetOS, IHost h
             IsDump = true;
             Architecture = dataReader.Architecture;
 
-            if (dataReader.ProcessId != -1) {
+            if (dataReader.ProcessId != -1)
+            {
                 ProcessId = (uint)dataReader.ProcessId;
             }
 
             OnFlushEvent.Register(dataReader.FlushCachedData);
 
             // Add the thread, memory, and module services
-            _serviceContainer.AddServiceFactory<IThreadService>((services) => new ThreadServiceFromDataReader(services, _dataReader));
-            _serviceContainer.AddServiceFactory<IModuleService>((services) => new ModuleServiceFromDataReader(services, _dataReader));
-            _serviceContainer.AddServiceFactory<IMemoryService>((_) => {
+            _serviceContainerFactory.AddServiceFactory<IThreadService>((services) => new ThreadServiceFromDataReader(services, _dataReader));
+            _serviceContainerFactory.AddServiceFactory<IModuleService>((services) => new ModuleServiceFromDataReader(services, _dataReader));
+            _serviceContainerFactory.AddServiceFactory<IMemoryService>((_) => {
                 IMemoryService memoryService = new MemoryServiceFromDataReader(_dataReader);
                 if (IsDump)
                 {
+                    ServiceContainerFactory clone = _serviceContainerFactory.Clone();
+                    clone.RemoveServiceFactory<IMemoryService>();
+
                     // The underlying host (dotnet-dump usually) doesn't map native modules into the address space
-                    memoryService = new ImageMappingMemoryService(_serviceContainer, memoryService, managed: false);
+                    memoryService = new ImageMappingMemoryService(clone.Build(), memoryService, managed: false);
```

Comment:
RemoveServiceFactory is too make sure the IMemoryService is the one the clone adds.  If the resulting service container built form the cloned service container factory flushes or removes the IMemoryService, it would default back to the wrong factory (the original one that all of this is happening in).

Cloning is absolutely vital here to give the image memory service a copy of the all the services using the passed in IMemoryService instance.  This removes any recursion when the image memory service reads memory directly or indirectly through other services like the IModuleService, etc.

Reply:
That clone issue warrants a comment. And I mostly meant if you need to call build twice. Or can the two memory services share containers.